### PR TITLE
Added support for Windows 64 bit MSYS2 Build (LP64)

### PR DESCRIPTION
**Pull request by spoiledsport**
Here it goes again !!!!!!!!!

Signed-off-by: Edison M. Castro <edisonmcastro@hotmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/311)

### CODE REVIEW

**Comment by tavip** [comment]
@lkl-jenkins: test this please

**Comment by tavip** [comment]
@Rondom can you please take a look?

File: tools/lkl/lib/nt-host.c [hunk 1]
```diff
@@ -288,6 +288,88 @@ struct lkl_host_operations lkl_host_ops = {
 	.jmp_buf_longjmp = jmp_buf_longjmp,
 };
 
+#ifdef PEX86_64
```
Comment (tavip) [inline tools/lkl/lib/nt-host.c]: Why is this required? Is it because of msys runs as posix? If so, shouldn't we use the full posix ops instead of nt?

Comment (Rondom) [inline tools/lkl/lib/nt-host.c]: Agreed. I also wondered why that should be necessary. We should use the POSIX for simplicity in maintainability. (One can use both APIs under Cygwin and MSYS, though)

**Comment by Rondom** [comment]
Overall, I think we really should not mix up MinGW with Cygwin/MSYS.
So I think we should check for MSYS/Cygwin separately to chose the default host-ops. The default executable output format returned by LD is not enough to determine this. Instead, we should probably use the triplet (gcc machinename) to determine the default host-ops. For MingW it would be NT, for Cygwin/MSYS2, POSIX should be the default (even though NT can also be used, if desired).

I had actually ported the cptofs, fs2tar and lklfuse to MINGW32 using GNULib. All I needed to do is change how the disk is opened and closed using ugly #ifdefs. This is what I would suggest instead of the changes to nt-host.

What about introducing two new functions that open and close the backing store of disks? The function signature might have to be generic which would make them clunky to use. Maybe just `open_disk(void*)` or `open_disk(char*, void*)`?

**Comment by tavip** [comment]
@Rondom I agree with you, we need to be more structured when dealing with lkl_disk. 

My proposal is a bit different. Instead of using an open function maybe we can use the same approach as for netdev where we have lkl_netdev_create_xxx that returns a struct lkl_netdev device which has a close method.

**Comment by spoiledsport** [comment]
You guys got the right idea
CYGWIN is ALMOST_POSIX
We can't use the regular POSIX_HOT file thought.
We need an abstraction to open and close the disk
All the fd (int) stuff should be delegated to the host code. There is a lot of that in the utilities (cptofs, etc)

**Comment by Rondom** [comment]
@tavip 
I agree, that sounds like a much better design and with the additional advantage that it is consistent with the netdevs. :-)

@spoiledsport 
> All the fd (int) stuff should be delegated to the host code. There is a lot of that in the utilities (cptofs, etc)

A lot? The only operations that need to be changed is opening and closing the disks., which is around 5 lines. As outlined above it is best to move this to dedicated functions.

> We can't use the regular POSIX_HOT file thought.

Can you please elaborate what you mean by POSIX_HOT and why?

**Comment by spoiledsport** [comment]
@Rondom
I meant to say POSIX_HOST file. Sorry for the crudeness of my response. The problem is that CYGWIN is not 100% Posix. It can't be. I initially went that way, but the compilation failed. I don't remember exactly, some missing or misplace include or something like that.

Also, I may have exaggerated the claim of "lots" of FD usage, but the point I wanted to make is that this knowledge leaks to the implementation of the toolset or utility. We should try, as much as possible, to keep this as a "black box" from implementers. 

Both
Your ideas are right on the money. An abstraction is needed to make this better. Let me know what needs to be done in order to get this patch thru

**Comment by tavip** [comment]
@spoiledsport What we are suggesting is to use either full posix or nt host ops. If there are reasons we can't support msys with full posix or nt ops, we should clearly document why.

Then there is the issue of creating disk APIs that is similar with lkl_netdev_create_* APIs, which IMO is an orthogonal issue that can be solved in a separate PR.

**Comment by spoiledsport** [comment]
I have 2 commits. One to enable 64-bit support in windows. That doesn't exists currently. The other is to enable those utilities to compile with the new support.

I don't think you have a problem with the first commit, do you? If so, can you get that that pulled in. I am still trying to figure out the changes required to make the 2nd commit palatable. I was really trying to avoid having a bunch of #ifdefs in those utilities to make them usable. 

@Rondom 

Can you push your changes that you mentioned above for supporting mingw 32 bit? I will make my changes afterwards to support 64 bit.

**Comment by spoiledsport** [comment]
I removed those changes that pertained to the utilities (cptofs, etc). I left just the changes to enable LP64 compilation in MYS2. Like I mentioned before this doesn't work currently. We get an unknown architecture error. I think also that the README file will have to be changes somehow.

**Comment by Rondom** [comment]
@spoiledsport As I said I was using GNUlib to make the cptofs, fs2tar and lklfuse run under MinGW32. This means that I had to use autotools to build the whole thing. Given the small differences, this is overkill and NOT the approach to take for a small utility that is inside the kernel-tree. This was a quick proof-of-concept. So my changes, are not applicable so easily.

Again, the differences are (from the top of my head):
* missing `stat` members and constants (either fixed using #ifdef or GNUlib stat, cannot remember)
* argp (fixed using GNUlib. libargp for Cygwin/MSYS2, but not for MinGW. I think there is no reason why libargp could not be made to work under MinGW with some patches to adapt for MinGW's getopt)
* opening files (simply open the images using `CreateFile` when using nt-host and `open` when using posix-host)

I will have a closer look at your PR now and try it again on a Windows-VM. The last time I could not get it to compile.
I use Windows only very seldomly. Is there anything specific you have done to your MSYS2-installation or while compiling or is a `git clone` followed by a  `make -C tools/lkl` enough? Have you tried compiling under Cygwin?

Note that MSYS2 is officially not "meant" to be used for production software, because they "do not want to duplicate Cygwin's efforts". The MSYS2-authors see MSYS2 as a (temporary) cygwin-fork providing a POSIXy environment for for compiling MINGW-applications. (See their wiki for more information).

**Comment by spoiledsport** [comment]
@Rondom

git clone followed by a make -C tools/lkl is enough. I didn't try Cygwin since I prefer MSYS :-). It is supposed to work equally. The main thing for me is the LP64 support since I am thinking about supporting 64bit file operations in some utilities I am working on.

I am looking also at your observation on disk operations. I will have something for that also real soon. Following that, I want to support NET operations. 

Let me know if you encounter any issues.

**Comment by Rondom** [comment]
Regarding LFS: On the lkl-side `lkl_off_t` is 64-bit. You only have to take care to `lkl_sys_open` with `O_LARGE_FILE`. I initially wanted to submit a PR to fix this in the utilities, but then I came across another patch that makes `O_LARGE_FILE` the default. This is the sensible thing to do, because lkl is a new architecture and there is not need to introduce legacy cruft into a new architecture. I only need to write some simple test to test the functionality.

On the host-side ` -D_FILE_OFFSET_BITS=64` should do the trick. This should be added to the `tools/lkl/Makefile`. I might as well change that in that PR as well.

>  I didn't try Cygwin since I prefer MSYS

So do I. In the long-term I hope that the projects merge. It was only a FYI, because I have seen MSYS2 being not very receptive of any patches that are more in the "use-MSYS-as-a-runtime-for-whatever-nice-utility-there-is-domain" and them stressing that over and over.

> I am looking also at your observation on disk operations. I will have something for that also real soon. 

Good to hear.

> Following that, I want to support NET operations.

Go for it :-D

Given that you solved some linker issues, I might have a try to look at solving the LLP64-issue on MINGW64 again in a few weeks time, because I actually do not really like Cygwin/MSYS2 at all :-P

**Comment by tavip** [comment]
@lkl-jenkins: test this please

**Comment by tavip** [comment]
@lkl-jenkins: retest this please

**Comment by tavip** [comment]
@spoiledsport I've squashed the commits and merge. For the future: when fixing a commit please don't add a new commit on top but update the commit directly. This helps us keeping the history clean.

**Comment by Rondom** [comment]
I am using latest MSYS2 and it fails during linking (writing from my mobile, can post the error later).

**Pull request merged**